HiP-DataStore/Core/WriteModel/HighScoreIndex.cs

```diff
@@ -0,0 +1,84 @@
+        /// Checks the in-memory index to detect whether there is already a recoreded highscore for the specified exhibit and user
+        /// </summary>
+        /// <param name="exhibitId">Exhibit ID</param>
+        /// <param name="userId">User's ID</param>
+        /// <returns>The entity Id of the highscore for the specified exhibit and user; null otherwise</returns>
+        public int? CheckHighScoreExistence(int exhibitId, string userId)
+        {
+            lock (_lockObject)
+            {
+                if (_highScoresDict.TryGetValue(userId, out var value))
+                {
+                    if (value.ExhibitIds.Contains(exhibitId))
+                    {
+                        int index = value.ExhibitIds.IndexOf(exhibitId);
+                        return value.EntityIds[index];
+                    }
+                }
+                return null;
+            }
+        }
+
+        /// <summary>
+        ///This implemented interface method is used to populate the in-memory highscore index to be used for validiation purposes when adding a highscore for specified exhibit and user
+        /// </summary>
+        /// <param name="e">Generic Event</param>
+        public void ApplyEvent(IEvent e)
+        {
+            switch (e)
+            {
+                case PropertyChangedEvent ev:
+                    if (ev.GetEntityType() == ResourceTypes.Highscore)
+                    {
+                        lock (_lockObject)
+                        {
+                            if (ev.PropertyName == nameof(ExhibitHighScoreArgs.ExhibitId) && ev.Value is int exhibitId)
+                            {
+                                HighScoreEntityInfo value;
+                                if (_highScoresDict.ContainsKey(ev.UserId))
+                                {
+                                    //update the entry in the dictionary
+                                    if (_highScoresDict.TryGetValue(ev.UserId, out value))
+                                    {
+                                        value.ExhibitIds.Add(exhibitId);
+                                        value.EntityIds.Add(ev.Id);
+                                        _highScoresDict[ev.UserId] = value;
+                                    }
+                                }
+                                else
+                                {
+                                    //create new pair where the key is the UserId
+                                    value = new HighScoreEntityInfo();
+                                    value.ExhibitIds.Add(exhibitId);
+                                    value.EntityIds.Add(ev.Id);
+                                    _highScoresDict.Add(ev.UserId, value);
+                                }
+                            }
+                        }
+                    }
+                    break;
+            }
+        }
+
+        private class HighScoreEntityInfo
```

Comment:
I would prefer if the HighscoreEntityInfo only contains one exhibitId and one entityId. The dictionary would then contain a list of HighScoreEntityInfo's instead of just one HighScoreEntityInfo

Reply:
So what is wrong exactly with the current implementation? would you please elaborate more on why I should change the current dictionary implementation? or it is just a personal preference!

Reply:
In my opinion it is easier to manage one list instead two. In the current way you always have to remember that you have to update both lists at the same time. Your implementation works but the code would be more maintainable with my suggested changes.